[#13894] [Docs] Add docs page for concurrency control + miscellaneous related changes (#14839)
yugabyte/yugabyte-db: docs/content/preview/architecture/transactions/read-committed.md

3. `BEGIN [TRANSACTION]; SET TRANSACTION ISOLATION LEVEL READ COMMITTED;` (this will be supported after [#12494](https://github.com/yugabyte/yugabyte-db/issues/12494))
4. `BEGIN [TRANSACTION]; SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL READ COMMITTED;` (this will be supported after #12494)

Semantics of Read Committed isolation adheres only with the [Wait-on-Conflict](../concurrency-control/#wait-on-conflict-beta-preview-faq-general-what-is-the-definition-of-the-beta-feature-tag) concurrency control policy. This is because a Read Committed transaction has to wait for other transactions to commit or rollback in case of a conflict, and then perform the re-check steps to make progress.

> As the [Fail-on-Conflict](../concurrency-control/#fail-on-conflict) concurrency control policy doesn't make sense for Read Committed, even if this policy is set for use on the cluster, transactions in Read Committed isolation provide `Wait-on-Conflict` semantics. This is done even without wait queues, that is, even when `enable_wait_queues=false`, by relying on an indefinite retry-backoff mechanism with exponential delays when conflicts are detected. However, when Read Committed isolation provides Wait-on-Conflict semantics without wait queues, the following limitations exist:

Author: Not able to follow how, can you elaborate?

> As the [Fail-on-Conflict](../concurrency-control/#fail-on-conflict) concurrency control policy doesn't make sense for Read Committed, even if this policy is set for use on the cluster, transactions in Read Committed isolation provide `Wait-on-Conflict` semantics. This is done even without wait queues, that is, even when `enable_wait_queues=false`, by relying on an indefinite retry-backoff mechanism with exponential delays when conflicts are detected. However, when Read Committed isolation provides Wait-on-Conflict semantics without wait queues, the following limitations exist:

Author: Tried finding a better way, but adding all information about behaviour with `enable_wait_queues=true` is making it hard to follow. Instead, I have just hyper-linked it and added the `has to wait for other transactions to commit or rollback in case of a conflict` phrase to give the crux of behaviour with wait queues.

Let me know of another way to structure this

> As the [Fail-on-Conflict](../concurrency-control/#fail-on-conflict) concurrency control policy doesn't make sense for Read Committed, even if this policy is set for use on the cluster, transactions in Read Committed isolation provide `Wait-on-Conflict` semantics. This is done even without wait queues, that is, even when `enable_wait_queues=false`, by relying on an indefinite retry-backoff mechanism with exponential delays when conflicts are detected. However, when Read Committed isolation provides Wait-on-Conflict semantics without wait queues, the following limitations exist:

Reviewer: ```suggestion
As the [Fail-on-Conflict](../concurrency-control/#fail-on-conflict) concurrency control policy doesn't make sense for Read Committed, YugabyteDB will provide [Wait-on-Conflict](../concurrency-control/#wait-on-conflict) semantics even when `enable_wait_queues=false` by relying on an indefinite retry-backoff mechanism with exponential delays when conflicts are detected. In this case, the following limitations exist:
```

Author: done, but slightly differently

Author: I see that you were trying to make it more concise, but left the redundancy as is to avoid confusion

> As the [Fail-on-Conflict](../concurrency-control/#fail-on-conflict) concurrency control policy doesn't make sense for Read Committed, even if this policy is set for use on the cluster, transactions in Read Committed isolation provide `Wait-on-Conflict` semantics. This is done even without wait queues, that is, even when `enable_wait_queues=false`, by relying on an indefinite retry-backoff mechanism with exponential delays when conflicts are detected. However, when Read Committed isolation provides Wait-on-Conflict semantics without wait queues, the following limitations exist:

Reviewer: i like your resolution. just added a suggestion to make it a bit more concise

- You may have to manually tune the exponential backoff parameters for performance as explained in [Performance tuning](../read-committed/#performance-tuning).
- The app may have to rely on statement timeouts to [avoid deadlocks](#avoid-deadlocks-in-read-committed-transactions).

> - There may be unfairness and high P99 latencies during contention due to the retry-backoff mechanism.

Author: yes that is true, rephrased it

> - There may be unfairness during contention due to the retry-backoff mechanism, resulting in high P99 latencies.

Reviewer: btw, it seems this whole section is restating the information present in 
<img width="855" alt="Screen Shot 2023-01-14 at 1 07 10 PM" src="https://user-images.githubusercontent.com/1457684/212488566-ce508445-14d2-4a38-b682-76846167440c.png">

does it make sense to just have one section link to the other instead of having redundant information in two places?

Author: nice point! changed

> 1. there can be unfairness and high P99 latencies during contention due to the retry-backoff mechanism.

Author: "unfairness" occurs in the following situation -

Assume multiple transactions T1, T2, ..., T100 start one after the other in the same order and try to take a lock. If T50 gets a lock before T5 for example, that is unfair.

### Avoid deadlocks in Read Committed transactions

> When wait queues are not enabled, that is, `enable_wait_queues=false`, configure a statement timeout to avoid deadlocks by setting the `statement_timeout` parameter in `ysql_pg_conf_csv` YB-TServer gflag on cluster startup.

Author: yes

You can set these parameters on a per-session basis, or in the `ysql_pg_conf_csv` YB-TServer gflag on cluster startup.

> If the [Wait-on-Conflict](../concurrency-control/#wait-on-conflict) concurrency control policy is enabled, there won't be a need to manually tune these parameters for performance. Statements will restart only when all conflicting transactions have committed or rolled back, instead of retrying with an exponential backoff.

Reviewer 2: I manually added an anchor link in concurrency-control.md so this link works right. It will continue to work right after we eventually remove the beta tag from that heading in concurrency-control.md, too.

> If the [Wait-on-Conflict](../concurrency-control/#wait-on-conflict-beta-preview-faq-general-what-is-the-definition-of-the-beta-feature-tag) concurrency control policy is enabled, there won't be a need to manually tune these parameters for performance. Statements will restart only when all conflicting transactions have committed or rolled back, instead of retrying with an exponential backoff.

Reviewer 2: Nope it's not. Drop the `-beta-preview-faq-general-what-is-the-definition-of-the-beta-feature-tag` :)

> If the [Wait-on-Conflict](../concurrency-control/#wait-on-conflict-beta-preview-faq-general-what-is-the-definition-of-the-beta-feature-tag) concurrency control policy is enabled, there won't be a need to manually tune these parameters for performance. Statements will restart only when all conflicting transactions have committed or rolled back, instead of retrying with an exponential backoff.

Author: no, it doesn't work without that. Works when I have the `-beta......` suffix
